File: src/com/example/javas/medium/M017_CourseSchedule.java
/*
MEDIUM PROBLEM (Simplified for 25 min):
Course Schedule (Detect Cycle in Directed Graph)

REQUIREMENTS:
- Given numCourses and list of prerequisite pairs [a, b] (must take b before a)
- Determine if you can finish all courses
- Return true if possible, false if there's a circular dependency

APPROACH:
- Build adjacency list graph from prerequisites
- Use DFS with state tracking to detect cycles
- States: unvisited (0), visiting (1), visited (2)

DATA STRUCTURES:
- List<List<Integer>>: adjacency list for graph
- int[]: state array for each node

TIME COMPLEXITY: O(V + E) where V = courses, E = prerequisites
SPACE COMPLEXITY: O(V + E)
*/

package com.example.javas.medium;

import java.util.*;

public class M017_CourseSchedule {
    
    public static boolean canFinish(int numCourses, int[][] prerequisites) {
        // TODO: Implement course schedule checker
        // 1. Build adjacency list graph
        // 2. Create state array (0 = unvisited, 1 = visiting, 2 = visited)
        // 3. For each course, if unvisited:
        //    - Run DFS to check for cycle
        //    - If cycle found, return false
        // 4. Return true if no cycles
        
        return false;
    }
    
    private static boolean hasCycle(int course, List<List<Integer>> graph, int[] state) {
        // TODO: Implement cycle detection DFS
        // 1. If state[course] == 1 (visiting), cycle detected, return true
        // 2. If state[course] == 2 (visited), already processed, return false
        // 3. Mark state[course] = 1 (visiting)
        // 4. For each neighbor in graph.get(course):
        //    - If hasCycle(neighbor, graph, state), return true
        // 5. Mark state[course] = 2 (visited)
        // 6. Return false (no cycle)
        
        return false;
    }
    
    // TEST CASES
    public static void main(String[] args) {
        // Test 1: 2 courses, [1,0] - need course 0 before 1
        int[][] prereq1 = {{1, 0}};
        System.out.println("Can finish 2 courses with prereq [[1,0]]: " + canFinish(2, prereq1));
        
        // Test 2: 2 courses, [1,0],[0,1] - circular dependency
        int[][] prereq2 = {{1, 0}, {0, 1}};
        System.out.println("Can finish 2 courses with prereq [[1,0],[0,1]]: " + canFinish(2, prereq2));
        
        // Test 3: 4 courses, linear dependency
        int[][] prereq3 = {{1, 0}, {2, 1}, {3, 2}};
        System.out.println("Can finish 4 courses with prereq [[1,0],[2,1],[3,2]]: " + canFinish(4, prereq3));
        
        // Test 4: No prerequisites
        int[][] prereq4 = {};
        System.out.println("Can finish 3 courses with no prereq: " + canFinish(3, prereq4));
    }
}

File: pythonX/medium/m017_course_schedule.py
"""
M017: Course Schedule (Detect Cycle in Directed Graph)

Summary:
Determine if all courses can be finished given prerequisites.

Requirements:
Given numCourses and prerequisite pairs, detect circular dependencies.

Approach:
Build adjacency list, use DFS with state tracking to detect cycles.

Data Structures:
- list[list[int]]: adjacency list
- list[int]: state array (0=unvisited, 1=visiting, 2=visited)

Big O Analysis:
Time Complexity: O(V + E)
Space Complexity: O(V + E)
"""


def can_finish(num_courses: int, prerequisites: list[list[int]]) -> bool:
    """
    TODO: Check if courses can be completed
    1. Build adjacency list graph
    2. Create state array (0=unvisited, 1=visiting, 2=visited)
    3. For each course, if unvisited:
       - Run DFS to check for cycle
       - If cycle found, return False
    4. Return True
    """
    pass


def has_cycle(course: int, graph: list[list[int]], state: list[int]) -> bool:
    """
    TODO: Detect cycle using DFS
    1. If state[course] == 1, cycle detected, return True
    2. If state[course] == 2, already visited, return False
    3. Mark state[course] = 1 (visiting)
    4. For each neighbor:
       - If has_cycle(neighbor), return True
    5. Mark state[course] = 2 (visited)
    6. Return False
    """
    pass


if __name__ == "__main__":
    # Test Cases
    
    print("Test 1: Simple prerequisite")
    prereq1 = [[1, 0]]
    print(f"Can finish 2 courses: {can_finish(2, prereq1)}")  # Expected: True
    
    print("\nTest 2: Circular dependency")
    prereq2 = [[1, 0], [0, 1]]
    print(f"Can finish 2 courses: {can_finish(2, prereq2)}")  # Expected: False
    
    print("\nTest 3: Linear dependency")
    prereq3 = [[1, 0], [2, 1], [3, 2]]
    print(f"Can finish 4 courses: {can_finish(4, prereq3)}")  # Expected: True
    
    print("\nTest 4: No prerequisites")
    prereq4 = []
    print(f"Can finish 3 courses: {can_finish(3, prereq4)}")  # Expected: True
